--- shim-12/lib/security_policy.c	2017-04-27 14:58:33.000000000 +0000
+++ shim-12-new/lib/security_policy.c	2017-06-28 21:52:58.310208313 +0000
@@ -55,14 +55,14 @@
 static EFI_SECURITY_FILE_AUTHENTICATION_STATE esfas = NULL;
 static EFI_SECURITY2_FILE_AUTHENTICATION es2fa = NULL;
 
-static EFI_STATUS thunk_security_policy_authentication(
+EFI_STATUS thunk_security_policy_authentication(
 	const EFI_SECURITY_PROTOCOL *This,
 	UINT32 AuthenticationStatus,
 	const EFI_DEVICE_PATH_PROTOCOL *DevicePath
 						       ) 
 __attribute__((unused));
 
-static EFI_STATUS thunk_security2_policy_authentication(
+EFI_STATUS thunk_security2_policy_authentication(
 	const EFI_SECURITY2_PROTOCOL *This,
 	const EFI_DEVICE_PATH_PROTOCOL *DevicePath,
 	VOID *FileBuffer,
@@ -276,11 +276,11 @@
 	 * If it fails, use security2_protocol == NULL as indicator */
 	uefi_call_wrapper(BS->LocateProtocol, 3,
 			  &SECURITY2_PROTOCOL_GUID, NULL,
-			  &security2_protocol);
+			  (void **)&security2_protocol);
 
 	status = uefi_call_wrapper(BS->LocateProtocol, 3,
 				   &SECURITY_PROTOCOL_GUID, NULL,
-				   &security_protocol);
+				   (void **)&security_protocol);
 	if (status != EFI_SUCCESS)
 		/* This one is mandatory, so there's a serious problem */
 		return status;
@@ -288,12 +288,12 @@
 	if (security2_protocol) {
 		es2fa = security2_protocol->FileAuthentication;
 		security2_protocol->FileAuthentication = 
-			thunk_security2_policy_authentication;
+			(EFI_SECURITY2_FILE_AUTHENTICATION)thunk_security2_policy_authentication;
 	}
 
 	esfas = security_protocol->FileAuthenticationState;
 	security_protocol->FileAuthenticationState =
-		thunk_security_policy_authentication;
+		(EFI_SECURITY_FILE_AUTHENTICATION_STATE)thunk_security_policy_authentication;
 
 	if (hook)
 		extra_check = hook;
@@ -311,7 +311,7 @@
 
 		status = uefi_call_wrapper(BS->LocateProtocol, 3,
 					   &SECURITY_PROTOCOL_GUID, NULL,
-					   &security_protocol);
+					   (void **)&security_protocol);
 
 		if (status != EFI_SUCCESS)
 			return status;
@@ -328,7 +328,7 @@
 
 		status = uefi_call_wrapper(BS->LocateProtocol, 3,
 					   &SECURITY2_PROTOCOL_GUID, NULL,
-					   &security2_protocol);
+					   (void **)&security2_protocol);
 
 		if (status != EFI_SUCCESS)
 			return status;
